CIP-meta-assets (ERC20-like assets)/README.md

```diff
@@ -0,0 +1,578 @@
+
+The purpose of this transaction is to transfer tokens from the sender's account to one or more receiver accounts. The transaction MUST ensure that the transfer is valid according to the states of each account involved if necessary.
+
+The transaction MUST call the `transferManagerObserver` contract with the `TransferRedeemer`. External contracts should look for this redeemer to obtain information regarding the transfer. 
+
+Additionally, any UTxOs spent from the `transferManager` contract MUST be spent with the `SingleTransferRedeemer` with `Transfer` constructor (index `0`).
+
+The transaction MUST include:
+
+1. One or more inputs from the sender's account in the `transferManager` with `SingleTransferRedeemer` for each UTxO spent.
+2. Outputs to the receives' accounts in the `transferManager` as specified in the `TransferRedeemer` of the transfer manager observer.
+3. The `TransferRedeemer` for the `transferManager` (as observer) contract.
+4. one `Account` reference input for each of the parties involved (sender + receivers).
+
+The information in the `TransferRedeemer` MUST be validated by the observer.
+
+That includes that the sum of the programmable tokens coming from the `transferManager` is stirctly equal to `totInputAmt`.
+
+There is one output for each element of the 2nd field (`outputs`) starting at the index specified by the 4th field (`firstOutputIndex`), in the same order.
+
+Each of the outputs going to the `transferManager` MUST have `TransferManagerDatum` **inline** datum, and MUST make sure the first field of the datum matches the token name of the NFT present on the respective reference input `Account`.
+
+
+## Rationale: how does this CIP achieve its goals?
+<!-- The rationale fleshes out the specification by describing what motivated the design and what led to particular design decisions. It SHOULD describe alternate designs considered and related work. The rationale SHOULD provide evidence of consensus within the community and discuss significant objections or concerns raised during the discussion.
+
+It MUST also explain how the proposal affects the backward compatibility of existing solutions when applicable. If the proposal responds to a CPS, the 'Rationale' section SHOULD explain how it addresses the CPS, and answer any questions that the CPS poses for potential solutions.
+-->
+
+The [first proposed implementation](https://github.com/cardano-foundation/CIPs/pull/444/commits/525ce39a89bde1ddb62e126e347828e3bf0feb58) (which we could informally refer as v0) was quite different by the one shown in this document
+
+Main differences were in the proposed:
+- [use of sorted merkle trees to prove uniqueness](https://github.com/cardano-foundation/CIPs/pull/444/commits/525ce39a89bde1ddb62e126e347828e3bf0feb58#diff-370b6563a47be474523d4f4dbfdf120c567c3c0135752afb61dc16c9a2de8d74R72) of an account during creation;
+- account credentials as asset name
```

Comment:
using the credentials for the name could not guarantee the unique association of an account to a set of utxos

a user could create as many accounts with the same name, with different states (or control  data, or config, whatever you like) which could represent a security vulnerability from the perspective of the spending logic

Reply:
Currently, the `transferManager` is parameterized by the `stateManager`. I haven't fully thought this through, but what if it wasn't and instead the `stateManager` hash was in the `transferManager` datum? It could be something like:
1. `accountManager` - a (new) minting policy where the token name is the user's credential. This enables the DApp entity to see all registered credentials.  This `accountManager` does *not* control any state; it just enables using the list of associated tokens as a list of registered users.
2. `stateManager` - the current `stateManager` except it is parameterized by the `accountManager` hash *and* the user's credential. So all users get their own `stateManager` minting policy for the programmable token. The DApp entity can see the credential in the registered list (tokens for `accountManager`) and can then lookup all associated `stateManager` tokens for that credential. Minting new `state` tokens requires referencing the `account` token for that credential.
3. `transferManager` - the current `transferManager` except it is not parameterized by anything. Instead, the `stateManager` hash is also included in the `transferManager` datum. The `accountManager` hash could also go in the datum. This contract does **not** mint the programmable tokens. Instead, they programmable tokens will be deposited into it.

This approach could make the `transferManager` general purpose so each user only needs one address for all programmable tokens. This approach also makes it easy for users to check their statuses since *all* tokens minted with a given `stateManager` (parameterized with their credential) belongs to them. In other words, they can find all of their state UTxOs using just their credential. Thoughts?

Reply:
> This approach could make the transferManager general purpose so each user only needs one address for all programmable tokens

I find it hard to believe that many tokens will adhere to the same spending rules

Reply:
> I find it hard to believe that many tokens will adhere to the same spending rules

I don't think they will, but the differences could possibly be encoded into the state UTxOs. Currently, the lack of generality is a deal breaker IMO. Do you expect users to have a separate address for every stablecoin, every NFT collection, every goverment issued DID, etc? Do you expect frontends to actually add support for possibly an infinite number of addresses? I don't. You might get a few addresses to be supported, but that means the frontends effectively act as gatekeepers for which tokens can be programmable since unsupported tokens effectively don't exist. Perhaps there can't be a universal `transferManager`, but perhaps 2 or 3 different `transferManagers` can cover most use cases.